src/Adapter/Order/Refund/OrderProductRemover.php

```diff
@@ -0,0 +1,265 @@
+            );
+        }
+    }
+
+    /**
+     * @param Order $order
+     * @param OrderDetail $orderDetail
+     * @param int $quantity
+     * @param float $productPriceTaxIncl
+     * @param float $productPriceTaxExcl
+     * @param float $shippingDiffTaxIncl
+     * @param float $shippingDiffTaxExcl
+     *
+     * @throws DeleteProductFromOrderException
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function updateOrderDetail(
+        Order $order,
+        OrderDetail $orderDetail,
+        int $quantity,
+        float $productPriceTaxIncl,
+        float $productPriceTaxExcl,
+        float $shippingDiffTaxIncl,
+        float $shippingDiffTaxExcl
+    ) {
+        $orderDetail->product_quantity -= (int) $quantity;
+        if ($orderDetail->product_quantity == 0) {
+            if (!$orderDetail->delete()) {
+                throw new DeleteProductFromOrderException('Could not delete order detail');
+            }
+            if (count($order->getProductsDetail()) == 0) {
+                $history = new OrderHistory();
+                $history->id_order = (int) $order->id;
+                $history->changeIdOrderState(Configuration::get('PS_OS_CANCELED'), $order);
+                if (!$history->addWithemail()) {
+                    // email failure must not block order update process
+                    $this->logger->warning(
+                        $this->translator->trans(
+                            'Order history email could not be sent, test your email configuration in the Advanced Parameters > E-mail section of your back office.',
+                            [],
+                            'Admin.Orderscustomers.Notification'
+                        )
+                    );
+                }
+            }
+
+            $order->update();
+        } else {
+            $orderDetail->total_price_tax_incl -= $productPriceTaxIncl;
+            $orderDetail->total_price_tax_excl -= $productPriceTaxExcl;
+            $orderDetail->total_shipping_price_tax_incl -= $shippingDiffTaxIncl;
+            $orderDetail->total_shipping_price_tax_excl -= $shippingDiffTaxExcl;
+        }
+    }
+
+    /**
+     * @param Order $order
+     * @param OrderDetail $orderDetail
+     * @param int $quantity
+     */
+    private function deleteCustomization(Order $order, OrderDetail $orderDetail, int $quantity)
+    {
+        if (!(int) $order->getCurrentState()) {
+            throw new DeleteCustomizedProductFromOrderException('Could not get a valid Order state before deletion');
+        }
+
+        if ($order->hasBeenDelivered()) {
+            return Db::getInstance()->execute('UPDATE `' . _DB_PREFIX_ . 'customization` SET `quantity_returned` = `quantity_returned` + ' . (int) $quantity . ' WHERE `id_customization` = ' . (int) $orderDetail->id_customization . ' AND `id_cart` = ' . (int) $order->id_cart . ' AND `id_product` = ' . (int) $orderDetail->product_id);
```

Comment:
Wouldn't it be simpler to use the object to update this data?

Reply:
This class is a wrapper for the legacy code that is actually being used today when you cancel a product. 

The scope of this PR is only to make email failure non blocking 😓